In [1]:
import openpyxl
wb = openpyxl.load_workbook('variables.xlsx')     # 開啟 Excel 檔案

names = wb.sheetnames    # 讀取 Excel 裡所有工作表名稱
s1 = wb['112']        

print(names)
# 印出 title ( 工作表名稱 )、max_row 最大列數、max_column 最大行數
print(s1.title, s1.max_row, s1.max_column)
print(s1['A1'].value)        # 取出 A1 的內容
print(s1.cell(4, 1).value)   # 取出 A4 的內容

['112', '111', '110', '109', '108', '107', '106', '105', '104', '103', '102', '101', '100', '99']
112 393 173
區域別
　三重區　


In [2]:
lst = [int(n) for n in names]
data_death = {}
data_marriage = {}

city = ["臺北市", "新北市", "基隆市", "新竹市", "桃園市", "新竹縣", "宜蘭縣", "桃園縣",
        "臺中市", "苗栗縣", "彰化縣", "南投縣", "雲林縣", "高雄市", "臺南市", 
        "嘉義市", "嘉義縣", "屏東縣", "澎湖縣", "花蓮縣", "臺東縣", "金門縣", "連江縣"]

for year in lst:
    for i in range(2,392):
        sheet = wb[str(year)]  
        clean_str = (sheet.cell(i, 1).value).replace("　", "")
        clean_str = (clean_str).replace("※", "")
        clean_str = (clean_str).replace(" ", "")
        if clean_str in city:
            temp_city = clean_str[:-1]
        else:
            if temp_city+clean_str[:-1] in data_death:
                data_death[temp_city+clean_str[:-1]].append(int(sheet.cell(i, 2).value))
                data_marriage[temp_city+clean_str[:-1]].append(int(sheet.cell(i, 3).value))
            else:
                data_death[temp_city+clean_str[:-1]] = [int(sheet.cell(i, 2).value)]
                data_marriage[temp_city+clean_str[:-1]] = [int(sheet.cell(i, 3).value)]
print(data_death)
#print(len(data_death))

# 桃園市
data_death['桃園桃園'] = [2925, 2918, 2536, 2286, 2347, 2234, 2144, 2108, 1951, 1980, 1826, 1804, 1833, 1608]
data_marriage['桃園桃園'] = [2793, 2790, 2551, 2608, 2730, 2860, 2948, 3015, 3027, 2732, 2789, 2741, 3177, 2540]

import pandas as pd
df_death = pd.DataFrame(data_death)
df_marriage = pd.DataFrame(data_marriage)
df_death.index = lst
df_marriage.index = lst

df_death = df_death.T
df_marriage = df_marriage.T
print(df_marriage)


{'新北板橋': [3844, 4101, 3646, 3163, 3317, 3168, 3080, 3093, 3026, 2863, 2779, 2784, 2730, 2519], '新北三重': [3095, 3198, 2915, 2581, 2587, 2477, 2505, 2430, 2223, 2267, 2323, 2148, 2096, 1919], '新北中和': [3185, 3237, 2821, 2651, 2740, 2525, 2674, 2528, 2442, 2388, 2202, 2209, 2158, 1984], '新北永和': [1688, 1769, 1625, 1412, 1488, 1413, 1426, 1456, 1361, 1413, 1229, 1279, 1247, 1159], '新北新莊': [2608, 2676, 2344, 2110, 2040, 1971, 1992, 1927, 1798, 1714, 1679, 1597, 1590, 1490], '新北新店': [2537, 2636, 2268, 2084, 2077, 2023, 1968, 1944, 1805, 1864, 1733, 1805, 1740, 1530], '新北樹林': [1205, 1286, 1079, 1021, 1017, 969, 971, 1004, 901, 948, 819, 777, 797, 745], '新北鶯歌': [668, 735, 615, 548, 574, 573, 568, 533, 535, 464, 475, 457, 429, 448], '新北三峽': [869, 881, 884, 759, 849, 787, 756, 739, 713, 712, 667, 664, 633, 601], '新北淡水': [1412, 1365, 1187, 1121, 1062, 1047, 986, 917, 869, 818, 820, 742, 718, 669], '新北汐止': [1599, 1604, 1384, 1188, 1298, 1264, 1173, 1184, 1144, 1042, 1049, 946, 946, 916], '新北瑞芳': [483

In [3]:
wb = openpyxl.load_workbook('population.xlsx')     # 開啟 Excel 檔案

In [4]:

data_population = data_death
for i in data_population: data_population[i] = []

city = ["臺北市", "新北市", "基隆市", "新竹市", "桃園市", "新竹縣", "宜蘭縣", "桃園縣",
        "臺中市", "苗栗縣", "彰化縣", "南投縣", "雲林縣", "高雄市", "臺南市", 
        "嘉義市", "嘉義縣", "屏東縣", "澎湖縣", "花蓮縣", "臺東縣", "金門縣", "連江縣", '總計', '臺灣省', '福建省']

for year in lst:
    for i in range(7,792):
        sheet = wb[str(year)]  
        clean_str = (sheet.cell(i, 2).value).replace("　", "")
        clean_str = (clean_str).replace("※", "")
        clean_str = (clean_str).replace(" ", "")
        if clean_str in city:
            temp_city = clean_str[:-1]
        else:
            if i < 401:
                #print(temp_city+clean_str[:-1])
                #print(data_population[temp_city+clean_str[:-1]])
                #print(int(sheet.cell(i, 3).value))
                data_population[temp_city+clean_str[:-1]].append(int(sheet.cell(i, 3).value))
            else:
                data_population[temp_city+clean_str[:-1]][-1] += (int(sheet.cell(i, 3).value))


print(data_population)
print(len(data_population))



{'新北板橋': [554247, 549572, 552713, 557114, 556897, 554742, 551480, 552285, 554236, 555914, 556920, 557440, 555335, 554596], '新北三重': [383108, 379825, 381731, 385328, 386336, 385826, 387484, 388581, 388447, 389325, 389813, 390090, 390421, 389968], '新北中和': [407273, 403109, 406140, 411214, 413069, 412486, 413590, 414266, 414304, 415226, 415742, 416499, 414939, 414356], '新北永和': [214771, 212170, 214967, 219364, 220595, 221098, 222585, 223999, 225409, 227267, 229062, 230768, 232386, 234536], '新北新莊': [423607, 419966, 421381, 422653, 420473, 417754, 416524, 414156, 413243, 411711, 409760, 407012, 404089, 402204], '新北新店': [304868, 298594, 299696, 303532, 303445, 302231, 302089, 301168, 300267, 299730, 299017, 297637, 296581, 296411], '新北樹林': [180658, 180067, 181903, 183470, 183926, 183946, 184149, 184400, 184249, 184167, 183407, 182012, 179788, 176077], '新北鶯歌': [89080, 88406, 88131, 87674, 86869, 86361, 86593, 87231, 87965, 89075, 88637, 88336, 87683, 86821], '新北三峽': [115608, 115274, 116019, 1169

In [5]:
import pandas as pd
df_population = pd.DataFrame(data_population)
df_population.index = lst
df_population = df_population.T
(df_population)


112     111     110     109     108     107     106     105     104  \
新北板橋  554247  549572  552713  557114  556897  554742  551480  552285  554236   
新北三重  383108  379825  381731  385328  386336  385826  387484  388581  388447   
新北中和  407273  403109  406140  411214  413069  412486  413590  414266  414304   
新北永和  214771  212170  214967  219364  220595  221098  222585  223999  225409   
新北新莊  423607  419966  421381  422653  420473  417754  416524  414156  413243   
...      ...     ...     ...     ...     ...     ...     ...     ...     ...   
金門烏坵     664     670     672     662     679     682     685     669     663   
連江南竿    7827    7797    7657    7638    7599    7663    7544    7382    7409   
連江北竿    3136    3181    2998    2717    2595    2425    2360    2344    2333   
連江莒光    1504    1471    1496    1489    1543    1618    1624    1562    1535   
連江東引    1572    1534    1494    1435    1352    1350    1352    1307    1270   

         103     102     101     100     99   
新北板橋  555914  556920  557440  555335  554596  
新北三重  389325  389813  390090  390421  389968  
新北中和  415226  415742  416499  414939  414356  
新北永和  227267  229062  230768  232386  234536  
新北新莊  411711  409760  407012  404089  402204  
...      ...     ...     ...     ...     ...  
金門烏坵     672     656     608     564     538  
連江南竿    7496    7288    6780    6000    5904  
連江北竿    2320    2268    2071    1864    1801  
連江莒光    1424    1438    1327    1163    1186  
連江東引    1266    1171    1132    1079    1053  

[368 rows x 14 columns]

In [6]:
# 插入廟宇變數
import requests
import xml.etree.ElementTree as ET
import json
from collections import defaultdict


# 下載 XML 文件
url = 'https://religion.moi.gov.tw/Report/temple.xml'
response = requests.get(url)
xml_data = response.content

# 解析 XML 文件
root = ET.fromstring(xml_data)

# 定義一個遞歸函數來將 XML 元素轉換為字典
def xml_to_dict(element):
    data_dict = {element.tag: {} if element.attrib else None}
    children = list(element)
    if children:
        dd = defaultdict(list)
        for dc in map(xml_to_dict, children):
            for k, v in dc.items():
                dd[k].append(v)
        data_dict = {element.tag: {k: v[0] if len(v) == 1 else v
                                   for k, v in dd.items()}}
    if element.attrib:
        data_dict[element.tag].update(('@' + k, v)
                                      for k, v in element.attrib.items())
    if element.text:
        text = element.text.strip()
        if children or element.attrib:
            if text:
                data_dict[element.tag]['#text'] = text
        else:
            data_dict[element.tag] = text
    return data_dict

# 將 XML 轉換為字典
xml_dict = xml_to_dict(root)

# 將字典轉換為 JSON
json_data = json.dumps(xml_dict, indent=4, ensure_ascii=False)

# 輸出 JSON
#print(json_data)

temple_data = json.loads(json_data)

print(type(temple_data))

print(type(temple_data['ArrayOfOpenData_3']))

<class 'dict'>
<class 'dict'>


In [7]:
print(type(temple_data['ArrayOfOpenData_3']["OpenData_3"]))

area_lst = []
for i in temple_data['ArrayOfOpenData_3']["OpenData_3"]:
    city = i['行政區'][:-1]
    area = i['地址'][3:]
    area = area.replace('鄉', ' ')
    area = area.replace('鎮', ' ')
    area = area.replace('市', ' ')
    area = area.replace('區', ' ')
    city += (area.split())[0]
    area_lst.append(city)

temple_dict = {}
for i in area_lst:
    if i not in temple_dict:
        temple_dict[i] =1
else: 
    temple_dict[i] += 1
    print (1)

print(temple_dict)

<class 'list'>
1
{'臺南白河': 1, '臺中后里': 1, '臺南麻豆': 1, '宜蘭冬山': 1, '臺中新社': 1, '花蓮瑞穗': 1, '新竹北': 1, '高雄梓官': 1, '彰化和美': 1, '雲林莿桐': 1, '宜蘭員山': 1, '高雄旗津': 1, '屏東東港': 1, '彰化埔鹽': 1, '雲林虎尾': 1, '南投鹿谷': 1, '高雄甲仙': 1, '臺中烏日': 1, '臺中太平': 1, '苗栗西湖': 1, '屏東南州': 1, '高雄路竹': 1, '南投埔里': 1, '基隆暖暖': 1, '臺北大同': 1, '花蓮花蓮': 1, '臺中霧峰': 1, '臺東鹿野': 1, '臺中豐原': 1, '嘉義鹿草': 1, '嘉義竹崎': 1, '南投集集': 1, '嘉義大林': 1, '高雄三民': 1, '臺南新化': 1, '嘉義東': 1, '屏東里港': 1, '宜蘭五結': 1, '臺南仁德': 1, '臺南南': 1, '高雄大寮': 1, '屏東枋山': 1, '高雄前': 1, '屏東枋寮': 1, '臺中大甲': 1, '桃園大溪': 1, '雲林元長': 1, '桃園中壢': 1, '苗栗大湖': 1, '新北平溪': 1, '臺南安南': 1, '彰化北斗': 1, '臺北內湖': 1, '雲林褒忠': 1, '嘉義水上': 1, '新北三芝': 1, '彰化大村': 1, '新北中和': 1, '臺中大安': 1, '花蓮玉里': 1, '新竹北埔': 1, '臺中外埔': 1, '雲林斗南': 1, '花蓮吉安': 1, '基隆七堵': 1, '彰化二林': 1, '臺南中西': 1, '雲林口湖': 1, '臺南七股': 1, '新北三峽': 1, '臺北中山': 1, '雲林土庫': 1, '臺南大內': 1, '基隆仁愛': 1, '臺南六甲': 1, '嘉義大埔': 1, '屏東九如': 1, '新北八里': 1, '新北五股': 1, '南投國姓': 1, '臺南山上': 1, '宜蘭大同': 1, '嘉義西': 1, '屏東佳冬': 1, '高雄大社': 1, '苗栗公館': 1, '高雄永安': 1, '臺中大里': 1, '基隆中正': 1, '桃園大園': 

In [8]:
import numpy as np
from selenium import webdriver
import pandas as pd
from selenium.webdriver.common.keys import Keys


In [9]:
browser = webdriver.Chrome('./chromedriver')
browser.get('https://zh.wikipedia.org/zh-tw/2022%E5%B9%B4%E4%B8%AD%E8%8F%AF%E6%B0%91%E5%9C%8B%E7%9B%B4%E8%BD%84%E5%B8%82%E8%AD%B0%E5%93%A1%E5%8F%8A%E7%B8%A3%E5%B8%82%E8%AD%B0%E5%93%A1%E9%81%B8%E8%88%89%E9%81%B8%E8%88%89%E5%8D%80%E6%8A%95%E7%A5%A8%E7%B5%90%E6%9E%9C%E5%88%97%E8%A1%A8')
browser.implicitly_wait(3)



elements = browser.find_elements('xpath', "//span[@class='mw-headline']")
rates = browser.find_elements('xpath', "//li[3]")
num = browser.find_elements('xpath', "//li[1]")
max_num = browser.find_elements('xpath', '//td[@align="center"]')

#print((elements))
elements = [i.text for i in elements]
num = [i.text for i in num]
max_num = [i.text for i in max_num]
rates = [i.text for i in rates]
print(rates)
print(num)
print(max_num)

browser.close()

['', '', '', '檢視歷史', '', '', '', '', '', '', '', '', '', '', '', '', '', '', '中國主題', '編', '投票數（投票率）：281,265（67.97%）', '投票數（投票率）：210,706（67.74%）', '投票數（投票率）：218,240（68.34%）', '投票數（投票率）：183,233（67.4%）', '投票數（投票率）：175,381（66.87%）', '投票數（投票率）：292,673（67.99%）', '投票數（投票率）：3,846（53.58%）', '投票數（投票率）：2,889（59.13%）', '投票數（投票率）：119,034（55.17%）', '投票數（投票率）：126,320（55.05%）', '投票數（投票率）：188,091（55.96%）', '投票數（投票率）：273,211（56.91%）', '投票數（投票率）：262,252（58.69%）', '投票數（投票率）：199,616（58.93%）', '投票數（投票率）：102,188（58.56%）', '投票數（投票率）：275,938（55.47%）', '投票數（投票率）：165,360（57.60%）', '投票數（投票率）：30,934（57.82%）', '投票數（投票率）：113,043（54.48%）', '投票數（投票率）：11,784（41.06%）', '投票數（投票率）：5,967（47.67%）', '投票數（投票率）：211,634（58.88%）', '投票數（投票率）：74,632（56.21%）', '投票數（投票率）：97,093（59.90%）', '投票數（投票率）：77,736（61.43%）', '投票數（投票率）：38,724（57.76%）', '投票數（投票率）：45,350（60.15%）', '投票數（投票率）：200,215（60.58%）', '投票數（投票率）：108,200（61.23%）', '投票數（投票率）：79,788（57.83%）', '投票數（投票率）：59,822（67.97%）', '投票數（投票率）：27,226（67.52%）', '投票數（投票率）：35,739（63.21%）', '投票數

In [10]:
#print(elements)
constituency_lst = []
count = 0
for i in elements:
    if '選舉區' in i:
        constituency_lst.append(i)
print(len(constituency_lst))
vr_lst = []
#print(rates)
for i in rates:
    if '投票率' in i:
        i = i[9:]
        i = (i).replace(",", ".")
        ind = (i.index('（'))
        i = i[ind+1:-2]
        vr_lst.append(round(eval(i)/100, 4))
print(vr_lst[0:5])
print(len(vr_lst))

num_lst = []
for i in num:
    if '應選出名額' in i:
        i = i[6:]
        i = i.split('席')[0]
        num_lst.append(int(i))

#print(len(num_lst))
max_num_lst = [0]

for i in max_num:
    try:
        if int(i)>max_num_lst[-1]:
            max_num_lst[-1] = int(i)
        else:
            max_num_lst.append(int(i))
    except:
        pass
#print(len(max_num_lst))

multi_lst = []
for i in range(215):
    multi_lst.append(round(max_num_lst[i]/num_lst[i], 4))

print(multi_lst)

215
[0.6797, 0.6774, 0.6834, 0.674, 0.6687]
215
[2.0833, 1.4444, 1.7778, 1.625, 2.25, 2.0769, 2.0, 3.0, 1.5, 2.0, 2.5714, 1.6667, 2.3333, 1.8333, 1.6667, 1.7, 1.2, 2.0, 1.5, 2.6667, 4.0, 2.25, 2.25, 2.2, 2.5, 1.5, 1.5, 2.0909, 2.0, 1.5, 1.6667, 2.0, 3.0, 1.75, 2.3333, 1.3333, 1.8, 1.8, 2.2, 2.0, 1.6, 1.75, 2.0, 1.6667, 2.3333, 2.2, 1.75, 2.3333, 2.5, 5.0, 3.0, 3.0, 1.5, 1.4, 2.0, 2.0, 1.8, 2.3333, 2.1429, 1.6667, 1.8333, 2.0, 1.6, 2.0, 3.0, 1.3333, 1.6667, 1.6, 2.2222, 2.6, 1.25, 1.8571, 1.6, 1.875, 2.125, 2.5, 3.0, 2.0, 3.0, 2.0, 2.4286, 2.0, 1.5, 1.0, 1.0, 1.8, 1.3333, 1.5, 3.0, 1.3333, 1.0, 2.0, 1.0, 1.8333, 2.0, 1.3333, 1.5, 1.5, 2.0, 2.0, 2.8333, 1.0, 3.0, 2.0, 2.0, 2.0, 1.7778, 1.6667, 1.6, 1.5556, 2.625, 2.0, 2.0, 2.0, 1.8462, 1.2857, 1.3333, 1.75, 1.4, 1.75, 1.6, 1.4, 2.0, 2.6, 1.5, 1.75, 1.4, 1.4286, 2.0, 2.0, 2.0, 1.8, 2.0, 1.5556, 1.1667, 1.5714, 1.4, 1.125, 1.7143, 1.8, 1.6667, 1.3333, 1.7143, 1.0, 2.0833, 1.5, 1.6, 1.625, 1.25, 2.3333, 3.0, 3.0, 2.0, 3.0, 1.0, 3.0, 2.0, 2.

In [11]:
count = 0
group_dic = {}
vr_dic = {}
for i, j, k in zip(constituency_lst, vr_lst, multi_lst):
    if '原住民' not in i:
        #print(i)
        city = i[0:2]
        ind = (i.index('（'))
        area = [_[:-1] for _ in i[ind+1:-1].split('、')]
        #print(area)
        for i in area:
            group_dic[city+i] = [count]
            vr_dic[count] = [j, 0, k, 0]
        count+=1
    
print(group_dic)
print(len(group_dic))
print(len(vr_dic))
print(vr_dic)



{'臺北北投': [0], '臺北士林': [0], '臺北內湖': [1], '臺北南港': [1], '臺北松山': [2], '臺北信義': [2], '臺北中山': [3], '臺北大同': [3], '臺北中正': [4], '臺北萬華': [4], '臺北大安': [5], '臺北文山': [5], '新北石門': [6], '新北三芝': [6], '新北淡水': [6], '新北八里': [6], '新北林口': [7], '新北五股': [7], '新北泰山': [7], '新北新莊': [8], '新北蘆洲': [9], '新北三重': [9], '新北板橋': [10], '新北中和': [11], '新北永和': [12], '新北樹林': [13], '新北鶯歌': [13], '新北土城': [13], '新北三峽': [13], '新北新店': [14], '新北深坑': [14], '新北石碇': [14], '新北坪林': [14], '新北烏來': [14], '新北平溪': [15], '新北瑞芳': [15], '新北雙溪': [15], '新北貢寮': [15], '新北金山': [16], '新北萬里': [16], '新北汐止': [16], '桃園桃園': [17], '桃園龜山': [18], '桃園八德': [19], '桃園蘆竹': [20], '桃園大園': [21], '桃園大溪': [22], '桃園復興': [22], '桃園中壢': [23], '桃園平鎮': [24], '桃園楊梅': [25], '桃園龍潭': [26], '桃園新屋': [27], '桃園觀音': [28], '臺中大甲': [29], '臺中大安': [29], '臺中外埔': [29], '臺中清水': [30], '臺中梧棲': [30], '臺中沙鹿': [30], '臺中龍井': [31], '臺中大肚': [31], '臺中烏日': [31], '臺中后里': [32], '臺中豐原': [32], '臺中神岡': [33], '臺中大雅': [33], '臺中潭子': [33], '臺中西屯': [34], '臺中南屯': [35], '臺中北屯': [36], '臺中北': [37], '臺中中': [38], '

In [12]:
df_index = df_death.index.tolist()

# 找出在 my_list 但不在 df_index 的元素
list_not_in_df = [item for item in group_dic if item not in df_index]
# 找出在 df_index 但不在 my_list 的元素
df_not_in_list = [item for item in df_index if item not in group_dic]

print(df_not_in_list)
print(list_not_in_df)

#新竹市的選區比較亂



['新竹東', '新竹北']
['新竹舊東區，東區東門里等35', '新竹舊南區，東區南門里等17', '新竹舊西區，北區客雅里等14', '新竹舊北區，北區北門里等30']


In [13]:
df_variable = pd.concat([df_death, df_marriage, df_population], axis=1)


df_variable.drop(df_not_in_list, inplace=True)
#print(df_variable)



for i in list_not_in_df:
    print(group_dic[i])
    del group_dic[i]
print(len(group_dic))

df_group = pd.DataFrame(group_dic)
df_group = df_group.T
df_group.rename(columns={'0': 'group'}, inplace=True)
df_group.head()

df_variable = pd.concat([df_variable, df_group], axis=1)
print(df_variable)



[146]
[147]
[148]
[149]
366
       112   111   110   109   108   107   106   105   104   103  ...     107  \
新北板橋  3844  4101  3646  3163  3317  3168  3080  3093  3026  2863  ...  554742   
新北三重  3095  3198  2915  2581  2587  2477  2505  2430  2223  2267  ...  385826   
新北中和  3185  3237  2821  2651  2740  2525  2674  2528  2442  2388  ...  412486   
新北永和  1688  1769  1625  1412  1488  1413  1426  1456  1361  1413  ...  221098   
新北新莊  2608  2676  2344  2110  2040  1971  1992  1927  1798  1714  ...  417754   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...     ...   
金門烏坵     5     4     3     2     2     2     3     4     6     3  ...     682   
連江南竿     0    45    41    39    36    32    34    37    32    41  ...    7663   
連江北竿    16    18    10    18    12    11    12     7     9    14  ...    2425   
連江莒光     8     8    12    12     7    12    13     4    10     8  ...    1618   
連江東引     4     9     4     4     3     8     7     5     2     2  ...    1350   


In [14]:
# 將字典中的值映射到DataFrame的索引，並將缺失值填充為0
df_variable['temple'] = df_variable.index.map(temple_dict).fillna(0)

# 如果需要將新列轉換為整數類型（可選）
df_variable['temple'] = df_variable['temple'].astype(int)

(df_variable)


112   111   110   109   108   107   106   105   104   103  ...     106  \
新北板橋  3844  4101  3646  3163  3317  3168  3080  3093  3026  2863  ...  551480   
新北三重  3095  3198  2915  2581  2587  2477  2505  2430  2223  2267  ...  387484   
新北中和  3185  3237  2821  2651  2740  2525  2674  2528  2442  2388  ...  413590   
新北永和  1688  1769  1625  1412  1488  1413  1426  1456  1361  1413  ...  222585   
新北新莊  2608  2676  2344  2110  2040  1971  1992  1927  1798  1714  ...  416524   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...     ...   
金門烏坵     5     4     3     2     2     2     3     4     6     3  ...     685   
連江南竿     0    45    41    39    36    32    34    37    32    41  ...    7544   
連江北竿    16    18    10    18    12    11    12     7     9    14  ...    2360   
連江莒光     8     8    12    12     7    12    13     4    10     8  ...    1624   
連江東引     4     9     4     4     3     8     7     5     2     2  ...    1352   

         105     104     103     102     101     100      99    0  temple  
新北板橋  552285  554236  555914  556920  557440  555335  554596   10       1  
新北三重  388581  388447  389325  389813  390090  390421  389968    9       1  
新北中和  414266  414304  415226  415742  416499  414939  414356   11       1  
新北永和  223999  225409  227267  229062  230768  232386  234536   12       1  
新北新莊  414156  413243  411711  409760  407012  404089  402204    8       1  
...      ...     ...     ...     ...     ...     ...     ...  ...     ...  
金門烏坵     669     663     672     656     608     564     538  153       0  
連江南竿    7382    7409    7496    7288    6780    6000    5904  156       1  
連江北竿    2344    2333    2320    2268    2071    1864    1801  157       1  
連江莒光    1562    1535    1424    1438    1327    1163    1186  158       1  
連江東引    1307    1270    1266    1171    1132    1079    1053  159       1  

[366 rows x 44 columns]

In [15]:
import numpy as np
grouped = df_variable.groupby(0).sum()
grouped.head()

grouped


112   111   110   109   108   107   106   105   104   103  ...     107  \
0                                                                ...           
0    4160  4264  3825  3532  3674  3695  3623  3545  3407  3390  ...  542076   
1    2626  2668  2291  2191  2212  2211  2128  2263  2147  2103  ...  409099   
2    3319  3286  2958  2750  2903  2925  2874  2963  2785  2707  ...  429108   
3    2866  3033  2662  2437  2549  2552  2523  2493  2407  2306  ...  357081   
4    2998  3290  3026  2738  2715  2814  2748  2868  2842  2804  ...  348603   
..    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...     ...   
155   112   102    89    78    84    81    80    74    58    58  ...   12861   
156     0    45    41    39    36    32    34    37    32    41  ...    7663   
157    16    18    10    18    12    11    12     7     9    14  ...    2425   
158     8     8    12    12     7    12    13     4    10     8  ...    1618   
159     4     9     4     4     3     8     7     5     2     2  ...    1350   

        106     105     104     103     102     101     100      99  temple  
0                                                                            
0    544751  547309  548556  547975  545186  543086  539362  533745       2  
1    409926  410249  409811  407024  402594  398515  391783  384268       2  
2    432741  436149  438925  440130  438839  438639  436489  433526       2  
3    359988  361318  362279  361469  358573  355468  350742  344726       2  
4    351458  353942  356814  358103  356620  354782  352304  348635       2  
..      ...     ...     ...     ...     ...     ...     ...     ...     ...  
155   12700   12568   12450   11978   11236   10424    9515    8921       1  
156    7544    7382    7409    7496    7288    6780    6000    5904       1  
157    2360    2344    2333    2320    2268    2071    1864    1801       1  
158    1624    1562    1535    1424    1438    1327    1163    1186       1  
159    1352    1307    1270    1266    1171    1132    1079    1053       1  

[156 rows x 43 columns]

In [16]:
#新增其他的投票率
browser = webdriver.Chrome('./chromedriver')
browser.get('https://zh.wikipedia.org/zh-tw/2018%E5%B9%B4%E4%B8%AD%E8%8F%AF%E6%B0%91%E5%9C%8B%E7%9B%B4%E8%BD%84%E5%B8%82%E8%AD%B0%E5%93%A1%E5%8F%8A%E7%B8%A3%E5%B8%82%E8%AD%B0%E5%93%A1%E9%81%B8%E8%88%89%E9%81%B8%E8%88%89%E5%8D%80%E6%8A%95%E7%A5%A8%E7%B5%90%E6%9E%9C%E5%88%97%E8%A1%A8#%E6%96%B0%E7%AB%B9%E5%B8%82%E7%AC%AC%E4%B8%80%E9%81%B8%E8%88%89%E5%8D%80%EF%BC%88%E8%88%8A%E6%9D%B1%E5%8D%80%EF%BC%8C%E6%9D%B1%E5%8D%80%E6%9D%B1%E9%96%80%E9%87%8C%E7%AD%8935%E9%87%8C%EF%BC%89')
browser.implicitly_wait(3)



elements = browser.find_elements('xpath', "//span[@class='mw-headline']")
rates = browser.find_elements('xpath', "//li[3]")
num = browser.find_elements('xpath', "//li[1]")
max_num = browser.find_elements('xpath', '//td[@align="center"]')

#print((elements))
elements = [i.text for i in elements]
num = [i.text for i in num]
max_num = [i.text for i in max_num]
rates = [i.text for i in rates]
browser.close()

#print(elements)
constituency_lst = []
count = 0
for i in elements:
    if '選舉區' in i:
        constituency_lst.append(i)
print(len(constituency_lst))
vr_lst = []
#print(rates)
for i in rates:
    if '投票率' in i:
        i = i[9:]
        i = (i).replace(",", ".")
        ind = (i.index('（'))
        i = i[ind+1:-2]
        vr_lst.append(round(eval(i)/100, 4))
print(vr_lst[0:5])
print(len(vr_lst))

num_lst = []
for i in num:
    if '應選出名額' in i:
        i = i[6:]
        i = i.split('席')[0]
        num_lst.append(int(i))

#print(len(num_lst))
max_num_lst = [0]

for i in max_num:
    try:
        if int(i)>max_num_lst[-1]:
            max_num_lst[-1] = int(i)
        else:
            max_num_lst.append(int(i))
    except:
        pass
#print(len(max_num_lst))

multi_lst = []
for i in range(len(max_num_lst)):
    multi_lst.append(round(max_num_lst[i]/num_lst[i], 4))

print(multi_lst)


count = 0
for i, j, k in zip(constituency_lst, vr_lst, multi_lst):
    if '原住民' not in i:
        #print(i)
        city = i[0:2]
        ind = (i.index('（'))
        area = [_[:-1] for _ in i[ind+1:-1].split('、')]
        #print(area)
        for i in area:
            if count == 9: #新莊區
                vr_dic[count][1] = (j)
                vr_dic[count+1][1] = (j)
                vr_dic[count][-1] = k
                vr_dic[count+1][-1] = k
                count+=1
                continue
            vr_dic[count][1] = (j)
            vr_dic[count][-1] = k
        count+=1
    






214
[0.6566, 0.668, 0.6607, 0.6573, 0.6588]
214
[1.7692, 1.6667, 2.1, 2.25, 2.375, 1.9231, 1.0, 2.0, 1.75, 2.1818, 1.6667, 2.0, 1.6667, 1.3333, 1.5, 1.6, 3.0, 2.0, 1.6667, 3.0, 1.8333, 1.5, 2.0, 1.75, 1.5, 2.5, 1.9091, 2.0, 2.5, 3.0, 3.0, 2.5, 2.25, 2.6667, 1.6667, 1.8, 1.8, 1.8, 1.6667, 1.6, 1.75, 2.6667, 2.3333, 2.3333, 2.2, 1.75, 1.6667, 2.0, 4.0, 2.0, 5.0, 1.6667, 2.0, 1.3333, 4.0, 2.0, 1.6667, 2.1429, 1.8333, 2.0, 2.2, 2.0, 3.0, 2.0, 2.0, 2.0, 1.4, 2.25, 2.0, 2.75, 1.5, 1.5, 2.0, 2.125, 2.25, 3.0, 4.0, 2.0, 2.0, 2.2727, 2.0, 1.3333, 1.5, 4.0, 2.0, 2.0, 2.1667, 4.0, 2.0, 2.0, 2.0, 4.0, 1.6667, 1.3333, 2.0, 1.7778, 2.125, 1.5, 4.0, 2.0, 1.6923, 1.4286, 1.3333, 1.625, 1.6, 1.75, 1.4, 1.8, 3.0, 1.6, 1.5, 1.5, 1.8, 1.4286, 2.0, 3.0, 3.0, 1.7, 1.6667, 2.125, 1.8333, 1.5714, 1.5, 1.5714, 1.2857, 1.6, 1.5714, 1.6667, 1.8571, 2.0, 1.6667, 1.625, 1.7, 1.25, 1.5, 2.0, 2.0, 4.0, 2.0, 2.0, 2.0, 1.0, 2.0, 6.0, 5.0, 4.0, 2.5714, 2.5, 2.0, 1.0, 1.5, 2.2, 2.3333, 2.0, 2.0, 1.6667, 2.0, 2.0, 4.0, 1

In [17]:
print(vr_dic)

{0: [0.6797, 0.6566, 2.0833, 1.7692], 1: [0.6774, 0.668, 1.4444, 1.6667], 2: [0.6834, 0.6607, 1.7778, 2.1], 3: [0.674, 0.6573, 1.625, 2.25], 4: [0.6687, 0.6588, 2.25, 2.375], 5: [0.6799, 0.6589, 2.0769, 1.9231], 6: [0.5517, 0.6245, 1.5, 1.75], 7: [0.5505, 0.6406, 2.0, 2.1818], 8: [0.5596, 0.633, 2.5714, 1.6667], 9: [0.5691, 0.6568, 1.6667, 2.0], 10: [0.5869, 0.6568, 2.3333, 2.0], 11: [0.5893, 0.6517, 1.8333, 1.6667], 12: [0.5856, 0.6619, 1.6667, 1.3333], 13: [0.5547, 0.6338, 1.7, 1.5], 14: [0.576, 0.6474, 1.2, 1.6], 15: [0.5782, 0.6642, 2.0, 3.0], 16: [0.5448, 0.6163, 1.5, 2.0], 17: [0.5888, 0.5727, 2.25, 1.8333], 18: [0.5621, 0.5856, 2.25, 1.5], 19: [0.599, 0.6052, 2.2, 2.0], 20: [0.6143, 0.5952, 2.5, 1.75], 21: [0.5776, 0.6513, 1.5, 1.5], 22: [0.6015, 0.6526, 1.5, 2.5], 23: [0.6058, 0.6098, 2.0909, 1.9091], 24: [0.6123, 0.6201, 2.0, 2.0], 25: [0.5783, 0.5986, 1.5, 2.5], 26: [0.6797, 0.6259, 1.6667, 3.0], 27: [0.6752, 0.7322, 2.0, 3.0], 28: [0.6321, 0.6599, 3.0, 2.5], 29: [0.6597, 0.7

In [18]:
df_voting= pd.DataFrame(vr_dic)

df_voting
df_voting = df_voting.T


df_voting.drop([146,147,148,149], inplace=True)

df_voting

0       1       2       3
0    0.6797  0.6566  2.0833  1.7692
1    0.6774  0.6680  1.4444  1.6667
2    0.6834  0.6607  1.7778  2.1000
3    0.6740  0.6573  1.6250  2.2500
4    0.6687  0.6588  2.2500  2.3750
..      ...     ...     ...     ...
155  0.3709  0.4204  1.5000  1.0000
156  0.7078  0.6992  1.6000  1.4000
157  0.7611  0.7468  1.5000  1.5000
158  0.7333  0.7809  2.0000  2.0000
159  0.7690  0.6429  2.0000  1.0000

[156 rows x 4 columns]

In [19]:

df_group = pd.concat([df_voting, grouped], axis=1)


df_group=df_group.T

df_group.index = range(1,48)
df_group=df_group.T
df_group

# 前兩項為投票率，而後每14項為循環 3-16 17-30 31-44
'''
112
2022 >> 111 110 109 108
2018 >> 107 106 105 104
2014 >> 103 102 101 100
99
'''
df_group['111d'] = round((df_group[6] + df_group[7] + df_group[8] + df_group[9]) / ((df_group[34] + df_group[35] + df_group[36] + df_group[37])/4), 4)
df_group['107d'] = round((df_group[10] + df_group[11] + df_group[12] + df_group[13]) / ((df_group[38] + df_group[39] + df_group[40] + df_group[41])/4), 4)

df_group['111m'] = round((df_group[20] + df_group[21] + df_group[22] + df_group[23]) / ((df_group[34] + df_group[35] + df_group[36] + df_group[37])/4), 4)
df_group['107m'] = round((df_group[24] + df_group[25] + df_group[26] + df_group[27]) / ((df_group[38] + df_group[39] + df_group[40] + df_group[41])/4), 4)

df_group['temple'] = df_group[47]

df_group['111d_1y'] = round((df_group[6]) / ((df_group[34])), 4)
df_group['107d_1y'] = round((df_group[10]) / ((df_group[38])), 4)


df_group['111m_1y'] = round((df_group[20]) / ((df_group[34])), 4)
df_group['107m_1y'] = round((df_group[24]) / ((df_group[38])), 4)

df_group['111d_'] = round((df_group[6] + df_group[7] + df_group[8] + df_group[9]))
df_group['107d_'] = round((df_group[10] + df_group[11] + df_group[12] + df_group[13]))

df_group['111m_'] = round((df_group[20] + df_group[21] + df_group[22] + df_group[23]))
df_group['107m_'] = round((df_group[24] + df_group[25] + df_group[26] + df_group[27]))

df_group['111p'] = ((df_group[34] + df_group[35] + df_group[36] + df_group[37])/4)
df_group['107p'] = ((df_group[38] + df_group[39] + df_group[40] + df_group[41])/4)


df_group

1       2       3       4       5       6       7       8       9  \
0    0.6797  0.6566  2.0833  1.7692  4160.0  4264.0  3825.0  3532.0  3674.0   
1    0.6774  0.6680  1.4444  1.6667  2626.0  2668.0  2291.0  2191.0  2212.0   
2    0.6834  0.6607  1.7778  2.1000  3319.0  3286.0  2958.0  2750.0  2903.0   
3    0.6740  0.6573  1.6250  2.2500  2866.0  3033.0  2662.0  2437.0  2549.0   
4    0.6687  0.6588  2.2500  2.3750  2998.0  3290.0  3026.0  2738.0  2715.0   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
155  0.3709  0.4204  1.5000  1.0000   112.0   102.0    89.0    78.0    84.0   
156  0.7078  0.6992  1.6000  1.4000     0.0    45.0    41.0    39.0    36.0   
157  0.7611  0.7468  1.5000  1.5000    16.0    18.0    10.0    18.0    12.0   
158  0.7333  0.7809  2.0000  2.0000     8.0     8.0    12.0    12.0     7.0   
159  0.7690  0.6429  2.0000  1.0000     4.0     9.0     4.0     4.0     3.0   

         10  ...  111d_1y  107d_1y  111m_1y  107m_1y    111d_    107d_  \
0    3695.0  ...   0.0084   0.0068   0.0057   0.0058  15295.0  14270.0   
1    2211.0  ...   0.0069   0.0054   0.0058   0.0060   9362.0   8749.0   
2    2925.0  ...   0.0083   0.0068   0.0055   0.0055  11897.0  11547.0   
3    2552.0  ...   0.0091   0.0071   0.0057   0.0060  10681.0   9975.0   
4    2814.0  ...   0.0102   0.0081   0.0051   0.0054  11769.0  11272.0   
..      ...  ...      ...      ...      ...      ...      ...      ...   
155    81.0  ...   0.0081   0.0063   0.0039   0.0030    353.0    293.0   
156    32.0  ...   0.0058   0.0042   0.0054   0.0070    161.0    135.0   
157    11.0  ...   0.0057   0.0045   0.0063   0.0066     58.0     39.0   
158    12.0  ...   0.0054   0.0074   0.0041   0.0019     39.0     39.0   
159     8.0  ...   0.0059   0.0059   0.0085   0.0111     20.0     22.0   

       111m_    107m_       111p       107p  
0    11027.0  13989.0  521311.50  545673.00  
1     8640.0  11155.0  396655.75  409771.25  
2     8247.0  10761.0  408594.25  434230.75  
3     7216.0   9634.0  342249.00  360166.50  
4     6482.0   8605.0  332994.75  352704.25  
..       ...      ...        ...        ...  
155    184.0    199.0   12718.50   12644.75  
156    157.0    176.0    7672.75    7499.50  
157     59.0     46.0    2872.75    2365.50  
158     30.0     28.0    1499.75    1584.75  
159     33.0     44.0    1453.75    1319.75  

[156 rows x 62 columns]

In [21]:

temp1 = df_group[[1, 3, 6, 20, '111d', '111m', 'temple','111d_', '111m_', '111p', '111d_1y', '111m_1y']]
temp1['vr'] = temp1[1]
temp1['multi'] = temp1[3]
temp1['year'] = 111
temp1['area'] = temp1.index
temp1['death'] = temp1['111d']
temp1['marriage'] = temp1['111m']
temp1['death_1y'] = temp1['111d_1y']
temp1['marriage_1y'] = temp1['111m_1y']
temp1['death_t'] = temp1['111d_']
temp1['marriage_t'] = temp1['111m_']
temp1['death_t1y'] = temp1[6]
temp1['marriage_t1y'] = temp1[20]
temp1['pop'] = temp1['111p']



temp2 = df_group[[2, 4, 10, 24,  '107d', '107m', 'temple','107d_', '107m_', '107p', '107d_1y', '107m_1y']]
temp2['vr'] = temp2[2]
temp2['multi'] = temp2[4]
temp2['year'] = 107
temp2['area'] = temp2.index
temp2['death'] = temp2['107d']
temp2['marriage'] = temp2['107m']
temp2['death_1y'] = temp2['107d_1y']
temp2['marriage_1y'] = temp2['107m_1y']
temp2['death_t'] = temp2['107d_']
temp2['marriage_t'] = temp2['107m_']
temp2['death_t1y'] = temp2[10]
temp2['marriage_t1y'] = temp2[24]
temp2['pop'] = temp2['107p']




<ipython-input-21-8c48d5397cbe>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp1['vr'] = temp1[1]
<ipython-input-21-8c48d5397cbe>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp1['multi'] = temp1[3]
<ipython-input-21-8c48d5397cbe>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning

In [24]:
df_reg = temp1.append(temp2, ignore_index=True)

df_reg

drop = [1,2, 3, 4, 6, 20, 10, 24, '111d', '111m', '111d_', '111m_', '111p', '107d', '107m', '107d_', '107m_', '107p','111d_1y', '111m_1y', '107d_1y', '107m_1y']
for i in drop: del df_reg[i]
(df_reg)

<ipython-input-24-f989528c66bc>:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_reg = temp1.append(temp2, ignore_index=True)


temple      vr   multi  year  area   death  marriage  death_1y  \
0       2.0  0.6797  2.0833   111     0  0.0293    0.0212    0.0084   
1       2.0  0.6774  1.4444   111     1  0.0236    0.0218    0.0069   
2       2.0  0.6834  1.7778   111     2  0.0291    0.0202    0.0083   
3       2.0  0.6740  1.6250   111     3  0.0312    0.0211    0.0091   
4       2.0  0.6687  2.2500   111     4  0.0353    0.0195    0.0102   
..      ...     ...     ...   ...   ...     ...       ...       ...   
307     1.0  0.4204  1.0000   107   155  0.0232    0.0157    0.0063   
308     1.0  0.6992  1.4000   107   156  0.0180    0.0235    0.0042   
309     1.0  0.7468  1.5000   107   157  0.0165    0.0194    0.0045   
310     1.0  0.7809  2.0000   107   158  0.0246    0.0177    0.0074   
311     1.0  0.6429  1.0000   107   159  0.0167    0.0333    0.0059   

     marriage_1y  death_t  marriage_t  death_t1y  marriage_t1y        pop  
0         0.0057  15295.0     11027.0     4264.0        2864.0  521311.50  
1         0.0058   9362.0      8640.0     2668.0        2226.0  396655.75  
2         0.0055  11897.0      8247.0     3286.0        2155.0  408594.25  
3         0.0057  10681.0      7216.0     3033.0        1880.0  342249.00  
4         0.0051  11769.0      6482.0     3290.0        1625.0  332994.75  
..           ...      ...         ...        ...           ...        ...  
307       0.0030    293.0       199.0       81.0          38.0   12644.75  
308       0.0070    135.0       176.0       32.0          54.0    7499.50  
309       0.0066     39.0        46.0       11.0          16.0    2365.50  
310       0.0019     39.0        28.0       12.0           3.0    1584.75  
311       0.0111     22.0        44.0        8.0          15.0    1319.75  

[312 rows x 14 columns]

$$turnover_i = \beta_0 + \beta_1 \ death_i + \beta_2 \ marriage_i + \alpha^{year}_i + \alpha^{region}_i +\delta \ temple_i +\beta X_i +ɛ_i$$

In [25]:
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf

model = smf.ols(formula='vr ~ death + marriage + C(year) + C(area) + temple + multi', data=df_reg)
results = model.fit()



print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                     vr   R-squared:                       0.844
Model:                            OLS   Adj. R-squared:                  0.681
Method:                 Least Squares   F-statistic:                     5.176
Date:                Sat, 22 Jun 2024   Prob (F-statistic):           1.88e-22
Time:                        22:40:40   Log-Likelihood:                 695.09
No. Observations:                 312   AIC:                            -1070.
Df Residuals:                     152   BIC:                            -471.3
Df Model:                         159                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          0.8078      0.087      9.

|        | coef   |  std err  |        t    |  P>abs(t)|
|  --- | ---  |  --- | ---  |  --- |
|death       |       1.5295  |    2.418   |   0.633   |   0.528     |
|marriage    |      -6.5511  |    2.753   |  -2.379   |   0.019     |
|temple      |       0.0015  |    0.010   |   0.152   |   0.879     |
|multi       |       0.0027  |    0.007   |   0.414   |   0.680     |

In [26]:
model = smf.ols(formula='vr ~ death_1y + marriage_1y + C(year) + C(area) + temple + multi', data=df_reg)
results = model.fit()



print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                     vr   R-squared:                       0.849
Model:                            OLS   Adj. R-squared:                  0.691
Method:                 Least Squares   F-statistic:                     5.381
Date:                Sat, 22 Jun 2024   Prob (F-statistic):           2.45e-23
Time:                        22:42:27   Log-Likelihood:                 700.22
No. Observations:                 312   AIC:                            -1080.
Df Residuals:                     152   BIC:                            -481.6
Df Model:                         159                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          0.7426      0.060     12.

| Variable      | Coef       | Std Err   | T         | P>abs(t) |
| ------------- | ---------- | --------- | --------- | -------- |
| death_1y      | 8.1568     | 4.989     | 1.635     | 0.104    |
| marriage_1y   | -18.9274   | 6.700     | -2.825    | 0.005    |
| temple        | -7.131e-05 | 0.006     | -0.012    | 0.991    |
| multi         | 0.0041     | 0.006     | 0.627     | 0.532    |


$$turnover_i = \beta_0 + \beta_1 \ death_i + \beta_2 \ marriage_i +\beta_3 \ population_i+ \alpha^{year}_i + \alpha^{region}_i + \beta X_i+ ɛ_i$$

In [30]:
model = smf.ols(formula='vr ~ death_t + marriage_t + pop + C(year) + C(area) + temple+multi', data=df_reg)
results = model.fit()

print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                     vr   R-squared:                       0.855
Model:                            OLS   Adj. R-squared:                  0.700
Method:                 Least Squares   F-statistic:                     5.545
Date:                Sat, 22 Jun 2024   Prob (F-statistic):           5.91e-24
Time:                        22:56:02   Log-Likelihood:                 705.92
No. Observations:                 312   AIC:                            -1090.
Df Residuals:                     151   BIC:                            -487.2
Df Model:                         160                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          1.7260      0.381      4.

| Variable       | Coef        | Std Err     | t Value   | P> abs(t)   |
| -------------- | ----------- | ----------- | --------- | ------- |
| death_t        | -7.304e-05  | 2.05e-05    | -3.565    | 0.000   |
| marriage_t     | -3.719e-05  | 1.39e-05    | -2.677    | 0.008   |
| pop            | 9.862e-07   | 9.44e-07    | 1.044     | 0.298   |
| temple         | -0.0158     | 0.020       | -0.800    | 0.425   |
| multi          | 0.0069      | 0.006       | 1.078     | 0.283   |


In [32]:
model = smf.ols(formula='vr ~ death_t1y + marriage_t1y + pop + C(year) + C(area) + temple+multi', data=df_reg)
results = model.fit()

print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                     vr   R-squared:                       0.856
Model:                            OLS   Adj. R-squared:                  0.704
Method:                 Least Squares   F-statistic:                     5.621
Date:                Sat, 22 Jun 2024   Prob (F-statistic):           2.87e-24
Time:                        22:58:04   Log-Likelihood:                 707.73
No. Observations:                 312   AIC:                            -1093.
Df Residuals:                     151   BIC:                            -490.8
Df Model:                         160                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept          1.8652      0.387      4.

|               | coef       | std err    | t         | P>abs(t)  |
|---------------|------------|------------|-----------|-----------|
| death_t1y     | -0.0001    | 2.97e-05   | -3.873    | 0.000     |
| marriage_t1y  | -0.0002    | 6.64e-05   | -3.075    | 0.002     |
| pop           | -7.184e-08 | 7.17e-07   | -0.100    | 0.920     |
| temple        | -0.0435    | 0.020      | -2.200    | 0.029     |
| multi         | 0.0081     | 0.006      | 1.265     | 0.208     |
